### PR TITLE
feat: refactor internal INTERVAL `Datum` implementation

### PR DESCRIPTION
**Pull request by alancai98**
## Relevant Issues
- Closes https://github.com/partiql/partiql-lang-kotlin/issues/1790
- Closes https://github.com/partiql/partiql-lang-kotlin/issues/1791

## Description
Refactors the internal `INTERVAL` representation. This new representation is less complicated than the prior implementation and ensures that the user does not need to perform normalization before using any Datum interval. See self-review comments for more details.

## Other Information
- Updated Unreleased Section in CHANGELOG: **[YES]**: <Explain if NO>
- Any backward-incompatible changes? **[NO]**: <Explain if YES>
- Any new external dependencies? **[NO]**: <Explain if YES>
- Do your changes comply with the [contributing][cg] and [code style][csg] guidelines? **[YES]**

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

[cg]: https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md
[csg]: https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md

### CODE REVIEW

File: partiql-spi/src/test/java/org/partiql/spi/value/__snapshots__/DatumToStringTest.snap [hunk 1]
```diff
@@ -129,7 +129,7 @@ DatumIntervalDayTime{_type=SECOND (2, 6), _value=INTERVAL '0 0:0:45.500000000'}
 
 
 org.partiql.spi.value.DatumToStringTest.testDatumIntervalYearMonthToString[datumMonth]=[
-DatumIntervalYearMonth{_type=MONTH (2), _value=INTERVAL '0-15'}
```
Comment (alancai98) [inline partiql-spi/src/test/java/org/partiql/spi/value/__snapshots__/DatumToStringTest.snap:132]: self-review: prior test should have normalized the month part of the year-month interval

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalYearMonth.java [hunk 2]
```diff
@@ -10,14 +10,18 @@
  */
 class DatumIntervalYearMonth implements Datum {
 
-    private final int _years;
-    private final int _months;
+    private final long _months;
```
Comment (alancai98) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalYearMonth.java:13]: self-review: for year-month intervals, only need to store the total number of months when doing computation. this simplifies the operations on intervals and removes the need for normalization.

Also considered using Java's `Period` https://docs.oracle.com/javase/8/docs/api/java/time/Period.html but it was slight overkill for just two fields.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java [hunk 3]
```diff
@@ -4,17 +4,20 @@
...
  * expect that all values have been normalized and checked already. This also lazily creates its type.
  */
 class DatumIntervalDayTime implements Datum {
 
-    private final int days;
-    private final int hours;
-    private final int minutes;
-    private final int seconds;
-    private final int nanos;
+    private final Duration _duration;
```
Comment (alancai98) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java:20]: self-review: using Java's `Duration` here which provides a lot of the datetime arithmetic operations that we need and performs its own normalization. The normalization slightly differs from what we want hence why there's some additional logic related to negative durations. E.g. Java's duration represents -1 nanoseconds as `-1 seconds` and `999999999 nanoseconds` what we really want is for all of the fields to be of the same sign (i.e. all positive or all negative).

File: partiql-spi/src/main/java/org/partiql/spi/value/Datum.java [hunk 4]
```diff
@@ -279,6 +279,42 @@ default int getMonths() throws InvalidOperationException, NullPointerException {
         throw new InvalidOperationException(getType(), "getMonths");
     }
 
+    /**
```
Comment (alancai98) [inline partiql-spi/src/main/java/org/partiql/spi/value/Datum.java:285]: self-review: marked as experimental for now. I'll see how the Scribe integration goes before making these non-experimental

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt [hunk 5]
```diff
@@ -71,38 +71,31 @@ internal object FnMinus : DiadicArithmeticOperator("minus") {
                 val p: Int = lhs.precision // TODO: Do we need to calculate a new precision?
                 val s: Int = 6 // TODO: Do we need to calculate a new fractional precision?
                 basic(PType.intervalDaySecond(p, s)) { args ->
-                    val interval0 = args[0]
-                    val interval1 = args[1]
-                    subtractIntervalDayTimes(interval0, interval1, p, s)
+                    val i1 = args[0]
```
Comment (alancai98) [inline partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt:75]: self-review: see here and in the other operator files for the simplified implementation of datetime arithmetic with the new representation

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt [hunk 6]
```diff
@@ -235,22 +238,25 @@ class DatumWriterTextPretty(
...
                 this.out.print(")")
             }
             IntervalCode.SECOND -> {
                 this.out.print("'")
-                this.out.print(datum.seconds)
-                this.out.print(fractionalSeconds(datum.nanos, type.fractionalPrecision))
+                if (datum.totalSeconds < 0) {
+                    this.out.print("-")
+                }
+                this.out.print((datum.days * SECONDS_PER_DAY + datum.hours * SECONDS_PER_HOUR + datum.minutes * SECONDS_PER_MINUTE + datum.seconds).absoluteValue)
```
Comment (xd1313113) [inline partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt]: Why do we need to do calculation before print? why not just `totalSeconds.absoluteValue`

File: partiql-spi/src/test/java/org/partiql/spi/value/__snapshots__/DatumToStringTest.snap (hunk 1 above)
Comment (xd1313113) [reply partiql-spi/src/test/java/org/partiql/spi/value/__snapshots__/DatumToStringTest.snap:132]: This is Month only interval. We should not normalize this one. Otherwise, it looks like a Year-Month interval.

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt [hunk 7]
```diff
@@ -288,7 +303,10 @@ class DatumWriterTextPretty(
             }
             IntervalCode.DAY_SECOND -> {
                 this.out.print("'")
-                this.out.print(datum.days)
+                if (datum.totalSeconds < 0 || (datum.totalSeconds == 0L && datum.nanos < 0)) {
```
Comment (XuechunHHH) [inline partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt]: nit: checking `datum.totalSeconds == 0L` seems unnecessary since `datum.nanos < 0` only when `datum.totalSeconds == 0L`.

File: partiql-cli/src/test/kotlin/org/partiql/cli/io/DatumWriterTextPrettyTests.kt [hunk 8]
```diff
@@ -146,6 +165,27 @@ class DatumWriterTextPrettyTests {
...
+            // Value greater than field divisor
+            Datum.intervalMonth(99, 2),
+            Datum.intervalHour(99, 2),
+            Datum.intervalMinute(99, 2),
+            Datum.intervalSecond(99, 0, 2, 2),
+            Datum.intervalSecond(-99, 1, 2, 2),
+            Datum.intervalMinuteSecond(99, 1, 0, 2, 2),
+            Datum.intervalHourSecond(99, 1, 2, 0, 2, 2),
+            Datum.intervalHourMinute(99, 1, 2),
+            Datum.intervalHourSecond(99, 1, 2, 3, 2, 2),
```
Comment (XuechunHHH) [inline partiql-cli/src/test/kotlin/org/partiql/cli/io/DatumWriterTextPrettyTests.kt:188]: Probably can add more tests for truncation.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java [hunk 9]
```diff
@@ -98,34 +98,58 @@ private void initType() {
...
     }
 
     @Override
     public int getHours() {
-        return hours;
+        return _duration.toHoursPart();
     }
 
     @Override
     public int getMinutes() throws InvalidOperationException, NullPointerException {
```
Comment (XuechunHHH) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java]: nit: looks unnecessary to throw error for `getMinutes()` and `getDays()`

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java [hunk 10]
```diff
@@ -98,34 +98,58 @@ private void initType() {
...
+        }
+        else {
+            return _duration.getSeconds();
+        }
     }
 
     @Override
     public String toString() {
+        String sign = "";
+        if (getTotalSeconds() == 0 && getNanos() < 0) {
```
Comment (XuechunHHH) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java]: Looks like should be `getTotalSeconds() < 0 || getNanos() < 0` here.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalYearMonth.java [hunk 11]
```diff
@@ -50,18 +54,27 @@ private void initType() {
...
+        return (int) (_months / 12);
     }
 
     @Override
     public int getMonths() throws InvalidOperationException, NullPointerException {
+        return (int) (_months % 12);
+    }
+
+    @Override
+    public long getTotalMonths() throws InvalidOperationException, NullPointerException {
```
Comment (XuechunHHH) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalYearMonth.java]: Same here. Seems will never throw an error.

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt [hunk 12]
```diff
@@ -71,38 +71,31 @@ internal object FnMinus : DiadicArithmeticOperator("minus") {
                 val p: Int = lhs.precision // TODO: Do we need to calculate a new precision?
                 val s: Int = 6 // TODO: Do we need to calculate a new fractional precision?
```
Comment (XuechunHHH) [inline partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt]: Why we pick 6 here instead of `maxOf(lhs.fractionalPrecision ?: 6, rhs.fractionalPrecision ?: 6)` and `lhs.precision` instead of `INTERVAL_MAX_PRECISION` (inconsistency with YM implementation)?

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt [hunk 13]
```diff
@@ -71,38 +71,31 @@ internal object FnMinus : DiadicArithmeticOperator("minus") {
...
                 basic(PType.intervalDaySecond(p, s)) { args ->
-                    val interval0 = args[0]
-                    val interval1 = args[1]
-                    subtractIntervalDayTimes(interval0, interval1, p, s)
+                    val i1 = args[0]
+                    val i2 = args[1]
+                    Datum.intervalDaySecond(
```
Comment (XuechunHHH) [inline partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt:77]: Shall we handle possible overflow cases here? e.g. ` INTERVAL '2147483647' DAY - INTERVAL '-1' DAY;`

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt [hunk 14]
```diff
@@ -71,38 +71,31 @@ internal object FnMinus : DiadicArithmeticOperator("minus") {
...
             lhs.code() == PType.INTERVAL_YM && rhs.code() == PType.INTERVAL_YM -> {
-                val p: Int = lhs.precision // TODO: Do we need to calculate a new precision?
+                val p = INTERVAL_MAX_PRECISION // Based on SQL2023 6.44 SR 2)c), precision is implementation-defined. Set to max
                 basic(PType.intervalYearMonth(p)) { args ->
-                    val interval0 = args[0]
-                    val interval1 = args[1]
-                    subtractIntervalYearMonths(interval0, interval1, p)
+                    val i1 = args[0]
+                    val i2 = args[1]
+                    Datum.intervalYearMonth(i1.years - i2.years, i1.months - i2.months, p)
```
Comment (XuechunHHH) [inline partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt:93]: Same here. Do we need to add some error handling?

**Review by XuechunHHH — COMMENTED** [review]
Nice refactor work on `INTERVAL`. With `duration`, the operator implementation looks much more cleaner and less error-prone. The code is significantly more readable and the logic flows better. Left comments at some places where I am a little confused but I may make some mistakes.

Also, are we planning to add some unit tests for these arithmetic or will we add those to conformance tests after finishing support for `INTERVAL?`

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt (hunk 7 above)
Comment (alancai98) [reply partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt]: Both branches are actually necessary. `totalSeconds` denotes the number of whole seconds (excluding nano seconds) while `nano` denotes the number of nanoseconds. We want to add the `-` sign whenever either
- total whole seconds is negative or
- when the total whole seconds is 0 but the number of nanoseconds is negative

For the first case, imagine we had `-10` whole seconds, the output should include the `-` sign (e.g. `INTERVAL '-10' SECOND`). The second case is a bit more subtle. Imagine we had `0` whole seconds and `-90000000` (e.g. `INTERVAL -0.9 SECOND`, ignoring any fractional precision).

Through the current representation of day-time intervals, there should not be a mix of signs between `totalSeconds` and `nano`.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java (hunk 10 above)
Comment (alancai98) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java]: Nice catch. ~I think it should replicate what we do in the Datum pretty printer, so `getTotalSeconds() < 0 || (getTotalSeconds() == 0 && getNanos() < 0)`.~

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java (hunk 9 above)
Comment (alancai98) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java]: Yeah, it's unnecessary. Copied over from the interface definition. I'll remove here and other places.

File: partiql-cli/src/test/kotlin/org/partiql/cli/io/DatumWriterTextPrettyTests.kt (hunk 8 above)
Comment (alancai98) [reply partiql-cli/src/test/kotlin/org/partiql/cli/io/DatumWriterTextPrettyTests.kt:188]: IMO, this isn't the right place to add since tests are for the CLI's datum pretty printer. Instead, I think we should add more truncation tests in the conformance tests (https://github.com/partiql/partiql-tests/issues/144). I'll add a list of more tests we should add in that issue.

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt (hunk 12 above)
Comment (alancai98) [reply partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt]: Ah, I forgot to update the day-time precision and fractional precision. I'll update.

Based on SQL2023 6.44, `precision` for leading field of year-month and day-time interval is implementation-defined, so I'll set it to `INTERVAL_MAX_PRECISION`. The fractional second precision is max of lhs and rhs fractional precision

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt (hunk 13 above)
Comment (alancai98) [reply partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt:77]: Overflow should not occur since the max precision of an interval field is `9`. The max number representable by an interval of precision `9` is less than `2147483647`, so overflow should not occur. There's already an existing function to check if an interval field exceeds the set precision https://github.com/partiql/partiql-lang-kotlin/blob/fd5493f616677aee9104c76c2c1dccaf331ad859/partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalHelpers.java#L20-L25.

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt (hunk 14 above)
Comment (alancai98) [reply partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt:93]: Similar to other comment for day-time interval subtraction. The max values are restricted by the max precision of `9`, which is less than the overflow threshold.

File: partiql-cli/src/test/kotlin/org/partiql/cli/io/DatumWriterTextPrettyTests.kt (hunk 8 above)
Comment (alancai98) [reply partiql-cli/src/test/kotlin/org/partiql/cli/io/DatumWriterTextPrettyTests.kt:188]: Added to the issue (https://github.com/partiql/partiql-tests/issues/144) description.

**Comment by alancai98** [comment]
> Also, are we planning to add some unit tests for these arithmetic or will we add those to conformance tests after finishing support for `INTERVAL?` 

We have existing unit tests in partiql-lang-kotlin for arithmetic +, -, *, / (see https://github.com/partiql/partiql-lang-kotlin/blob/fd5493f616677aee9104c76c2c1dccaf331ad859/partiql-eval/src/test/kotlin/org/partiql/eval/internal/IntervalPlusTests.kt for example). There very well could be some edge cases we aren't covering. Since we're going to migrate the evaluation tests to the conformance tests in the near future, I think we should track any `INTERVAL` testing gaps in https://github.com/partiql/partiql-tests/issues/144.

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt (hunk 7 above)
Comment (XuechunHHH) [reply partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt]: Emm, since there should not be a mix of sings between `totalSeconds` and `nano`, I think checking `datum.totalSeconds < 0 || datum.nanos < 0` is enough here. Both cases you mentioned above and a possible third case `-10.9` (`-10` whole seconds and `-90000000`) could be all included here.

File: partiql-cli/src/test/kotlin/org/partiql/cli/io/DatumWriterTextPrettyTests.kt (hunk 8 above)
Comment (XuechunHHH) [reply partiql-cli/src/test/kotlin/org/partiql/cli/io/DatumWriterTextPrettyTests.kt:188]: Ah yes, it would be more appropriate to be added in conformance tests.

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt [hunk 15]
```diff
@@ -235,22 +238,25 @@ class DatumWriterTextPretty(
             }
             IntervalCode.HOUR -> {
                 this.out.print("'")
-                this.out.print(datum.hours)
+                this.out.print(datum.totalSeconds / SECONDS_PER_HOUR)
```
Comment (xd1313113) [inline partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt:240]: I think we should keep months, hours, minutes etc as normalized value. Private setter and public getter. 
It is a bit odd here to calculate the value by caller itself.

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt (hunk 12 above)
Comment (XuechunHHH) [reply partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt]: Nice. That makes sense.

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt (hunk 13 above)
Comment (XuechunHHH) [reply partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnMinus.kt:77]: I see. Thanks for clarification.

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt [hunk 16]
```diff
@@ -235,22 +238,25 @@ class DatumWriterTextPretty(
...
+                this.out.print(datum.totalSeconds / SECONDS_PER_MINUTE)
                 this.out.print("' MINUTE (")
                 this.out.print(type.precision)
                 this.out.print(")")
             }
             IntervalCode.SECOND -> {
                 this.out.print("'")
-                this.out.print(datum.seconds)
-                this.out.print(fractionalSeconds(datum.nanos, type.fractionalPrecision))
+                if (datum.totalSeconds < 0) {
```
Comment (XuechunHHH) [inline partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt]: Seems to be `datum.totalSeconds < 0 || datum.nanos < 0)` here.

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt (hunk 6 above)
Comment (XuechunHHH) [inline partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt]: nit: use `datum.totalSeconds.absoluteValue` instead

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalYearMonth.java [hunk 17]
```diff
@@ -49,19 +53,28 @@ private void initType() {
     }
 
     @Override
-    public int getYears() throws InvalidOperationException, NullPointerException {
-        return _years;
+    public int getYears() {
+        return (int) (_months / 12);
```
Comment (xd1313113) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalYearMonth.java:57]: What should we return for getyears if it is Month interval or year-Month interval respectively?
E.g.  For 16 months,  if it is interval Month, the output should be "0-16" and if it is Interval YearMonth, we should output  "1-4", which is 1 year and 4 months.

File: partiql-cli/src/test/kotlin/org/partiql/cli/io/DatumWriterTextPrettyTests.kt [hunk 18]
```diff
@@ -74,6 +74,25 @@ class DatumWriterTextPrettyTests {
           INTERVAL '-1:2.30' MINUTE (2) TO SECOND (2),
           INTERVAL '3:4.00' MINUTE (2) TO SECOND (2),
           INTERVAL '3:4.012345678' MINUTE (2) TO SECOND (9),
```
Comment (XuechunHHH) [inline partiql-cli/src/test/kotlin/org/partiql/cli/io/DatumWriterTextPrettyTests.kt:76]: Can probably add more zero leading tests like: `INTERVAL '-0.10' SECOND (2,2)` accordingly here.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java [hunk 19]
```diff
@@ -97,35 +97,59 @@ private void initType() {
     }
 
     @Override
-    public int getDays() throws InvalidOperationException, NullPointerException {
-        return days;
+    public int getDays() {
```
Comment (xd1313113) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java:100]: Not sure if getDays should return different result based on interval type. And writeDatum should use those properties for print rather than calculating on its owns.

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnAbs.kt [hunk 20]
```diff
@@ -118,22 +117,11 @@ internal val Fn_ABS__INTERVAL_YM__INTERVAL_YM = Function.overload(
     parameters = arrayOf(Parameter("value", PType.intervalYearMonth(2))),
```
Comment (XuechunHHH) [inline partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnAbs.kt]: Also use `INTERVAL_MAX_PRECISION` instead here

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnAbs.kt [hunk 21]
```diff
@@ -142,66 +130,13 @@ internal val Fn_ABS__INTERVAL_DT__INTERVAL_DT = Function.overload(
     parameters = arrayOf(Parameter("value", PType.intervalDaySecond(2, 6))),
```
Comment (XuechunHHH) [inline partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnAbs.kt]: Also use `INTERVAL_MAX_PRECISION` and `NANO_MAX_PRECISION` instead here

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt (hunk 6 above)
Comment (XuechunHHH) [reply partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt]: Agree to use `datum.totalSeconds.absoluteValue` to simplify the code. Still doing calculation at print time tho.

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt (hunk 15 above)
Comment (XuechunHHH) [reply partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt:240]: > This is Month only interval. We should not normalize this one. Otherwise, it looks like a Year-Month interval.

`INTERVAL '0-15' YEAR TO MONTH` is not allowed now.

Comment (XuechunHHH) [reply partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt:240]: > I think we should keep months, hours, minutes etc as normalized value. Private setter and public getter.
> It is a bit odd here to calculate the value by caller itself.

We are currently using `Duration` as internal representation for `INTERVAL` and don't maintain months, hours, minutes separately any more.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalYearMonth.java (hunk 17 above)
Comment (XuechunHHH) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalYearMonth.java:57]: I think the `getYears()` is designed for getting `years` in `_months`. Can refer to:
> self-review: for year-month intervals, only need to store the total number of months when doing computation. this simplifies the operations on intervals and removes the need for normalization.

In the example you mentioned, it is used to get the `1 year`.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java (hunk 19 above)
Comment (XuechunHHH) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java:100]: Same for `getDays()` here. It is for getting `days` in `_duration`.

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt (hunk 7 above)
Comment (alancai98) [reply partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt]: Ah my bad, you're right. In a previous version of this refactor, I had modeled `totalSeconds` and `nano` more similarly to Java's `Duration`, which _can_ have mixed signs (e.g. a negative `totalSeconds` and a positive `nano` field). I will revise to your suggestion.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java (hunk 10 above)
Comment (alancai98) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java]: I'll revise to your suggestion.

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt (hunk 16 above)
Comment (alancai98) [reply partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt]: Fixed.

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt (hunk 6 above)
Comment (alancai98) [reply partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt]: My bad, this was some leftover logic from the prior implementation I was trying out. We can just use `datum.totalSeconds.absoluteValue`.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalYearMonth.java (hunk 17 above)
Comment (alancai98) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalYearMonth.java:57]: > What should we return for getyears if it is Month interval or year-Month interval respectively?
E.g. For 16 months, if it is interval Month, the output should be "0-16" and if it is Interval YearMonth, we should output "1-4", which is 1 year and 4 months.

It should be from the normalized value. For both of the above it will output `1` year.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java (hunk 19 above)
Comment (alancai98) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java:100]: > Not sure if getDays should return different result based on interval type. And writeDatum should use those properties for print rather than calculating on its owns.

Leaving the values unnormalized may make the printing slightly easier but it will make any other computation more complicated since a user of the API would need to check the interval type. IMO, when it comes to computation, all we really care about is the normalized value. E.g. `INTERVAL '2' YEAR` is effectively equivalent to `INTERVAL '24' MONTH`.

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt (hunk 15 above)
Comment (alancai98) [reply partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt:240]: > I think we should keep months, hours, minutes etc as normalized value. Private setter and public getter.
It is a bit odd here to calculate the value by caller itself.

What do you mean by normalized value? In the other comments you were suggesting not using the normalized value.

File: partiql-spi/src/test/java/org/partiql/spi/value/__snapshots__/DatumToStringTest.snap (hunk 1 above)
Comment (alancai98) [reply partiql-spi/src/test/java/org/partiql/spi/value/__snapshots__/DatumToStringTest.snap:132]: In the `.toString()` implementation, we already print out the interval type (i.e. that it's a month interval).

File: partiql-cli/src/test/kotlin/org/partiql/cli/io/DatumWriterTextPrettyTests.kt (hunk 18 above)
Comment (alancai98) [reply partiql-cli/src/test/kotlin/org/partiql/cli/io/DatumWriterTextPrettyTests.kt:76]: I already added some leading zero tests. See https://github.com/partiql/partiql-lang-kotlin/blob/1068fffb48581f87e0c53d3dc93b74a9da7b9576/partiql-cli/src/test/kotlin/org/partiql/cli/io/DatumWriterTextPrettyTests.kt#L77-L86

File: partiql-spi/src/test/java/org/partiql/spi/value/__snapshots__/DatumToStringTest.snap (hunk 1 above)
Comment (xd1313113) [reply partiql-spi/src/test/java/org/partiql/spi/value/__snapshots__/DatumToStringTest.snap:132]: Discussed offline, we will normalize the getter values. And remind users to pull full format to get accurate value.

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt [hunk 22]
```diff
@@ -327,7 +350,10 @@ class DatumWriterTextPretty(
             }
             IntervalCode.MINUTE_SECOND -> {
                 this.out.print("'")
-                this.out.print(datum.minutes)
+                if (datum.totalSeconds < 0 || datum.nanos < 0) {
```
Comment (xd1313113) [inline partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt:353]: The negative condition seems counter-intuitive due to java implementation, do we need expose an API in datum?

File: partiql-spi/src/main/java/org/partiql/spi/value/Datum.java [hunk 23]
```diff
@@ -279,6 +279,42 @@ default int getMonths() throws InvalidOperationException, NullPointerException {
...
+     * Returns the total amount of whole seconds for this day-time interval.
+     * </p>
+     * <p>
+     * This only applies to {@link PType#INTERVAL_DT}.
+     * </p>
+     * @return the total number of whole seconds.
+     * @throws InvalidOperationException when the type is not {@link PType#INTERVAL_DT}.
+     * @throws NullPointerException when {@link #isNull()} is true.
+     */
+    default long getTotalSeconds() throws InvalidOperationException, NullPointerException {
```
Comment (xd1313113) [inline partiql-spi/src/main/java/org/partiql/spi/value/Datum.java:318]: getTotalSeconds does not return total duration. Do you expect user to call `datum.nano` as well?

https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html#:~:text=The%20total%20duration%20is%20defined%20by%20calling%20this%20method%20and%20getNano().

File: partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt (hunk 22 above)
Comment (alancai98) [reply partiql-cli/src/main/kotlin/org/partiql/cli/io/DatumWriterTextPretty.kt:353]: IMO it's not counterintuitive. I'd rather not expose more functions in the public Datum API unless we need to.

File: partiql-spi/src/main/java/org/partiql/spi/value/Datum.java (hunk 23 above)
Comment (alancai98) [reply partiql-spi/src/main/java/org/partiql/spi/value/Datum.java:318]: Yes that's the expectation. It's included in the javadoc already that this returns just the whole seconds:

> Returns the total amount of whole seconds for this day-time interval.

Added to the javadoc about `nanos` in [`baec7f4` (#1815)](https://github.com/partiql/partiql-lang-kotlin/pull/1815/commits/baec7f47a19090098baf49eae21977016e55e325)

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumComparator.java [hunk 24]
```diff
@@ -558,24 +558,12 @@ private static DatumComparison[] fillTimeComparator(DatumComparison[] comps) {
...
-            long totalMonths = self.getYears() *  (long) 12 + self.getMonths();
-            long otherTotalMonths = other.getYears() * (long) 12 + other.getMonths();
-            return Long.compare(totalMonths, otherTotalMonths);
-        };
+        comps[INTERVAL_YM] = (self, other, comp) -> Long.compare(self.getTotalMonths(), self.getTotalMonths());
         comps[INTERVAL_DT] = (self, other, comp) -> {
-            long totalSeconds = self.getDays() * SECONDS_PER_DAY + self.getHours() * SECONDS_PER_HOUR + self.getMinutes() * SECONDS_PER_MINUTE + self.getSeconds();
-            long otherTotalSeconds = other.getDays() * SECONDS_PER_DAY + other.getHours() * SECONDS_PER_HOUR + other.getMinutes() * SECONDS_PER_MINUTE + other.getSeconds();
-            int comparison = Long.compare(totalSeconds, otherTotalSeconds);
+            int comparison = Long.compare(self.getTotalSeconds(), other.getTotalSeconds());
```
Comment (alancai98) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumComparator.java:566]: self-review: this was another place we can simplify the computation between intervals
